encoding/m3tsz/encoder.go

```diff
@@ -273,6 +357,108 @@ func (enc *encoder) writeXOR(prevXOR, curXOR uint64) {
 	enc.os.WriteBits(curXOR>>uint(curTrailing), numMeaningfulBits)
 }
 
+// writeIntVal writes the val as a diff of ints
+func (enc *encoder) writeIntVal(val float64, mult uint8, isFloat bool) {
+	valDiff := enc.intVal - val
+	if valDiff == 0 && isFloat == enc.isFloat {
+		// Value is repeated
+		enc.os.WriteBit(opcodeUpdate)
+		enc.os.WriteBit(opcodeRepeat)
+		return
+	}
+
+	neg := false
+	if valDiff < 0 {
+		neg = true
+		valDiff = -1 * valDiff
+	}
+
+	valDiffBits := uint64(int64(valDiff))
+	numSig := encoding.NumSig(valDiffBits)
+	newSig := enc.getNewSig(numSig)
+	if mult > enc.maxMult || enc.numSig != newSig || isFloat != enc.isFloat {
+
+		enc.os.WriteBit(opcodeUpdate)
+		enc.os.WriteBit(opcodeNoRepeat)
+		enc.os.WriteBit(opcodeIntMode)
+		enc.writeIntSigMult(newSig, mult)
+		enc.writeIntValDiff(valDiffBits, neg)
+		enc.isFloat = false
+	} else {
+		enc.os.WriteBit(opcodeNoUpdate)
+		enc.writeIntValDiff(valDiffBits, neg)
+	}
+
+	enc.intVal = val
+}
+
+// writeIntValDiff writes the provided val diff bits along with
+// whether the bits are negative or not
+func (enc *encoder) writeIntValDiff(valBits uint64, neg bool) {
+	if neg {
+		enc.os.WriteBit(opcodeNegative)
+	} else {
+		enc.os.WriteBit(opcodePositive)
+	}
+
+	enc.os.WriteBits(valBits, int(enc.numSig))
+}
+
+// writeIntSigMult writes the number of significant
+// bits of the diff and the multiplier if they have changed
+func (enc *encoder) writeIntSigMult(sig, mult uint8) {
+	if enc.numSig != sig {
+		enc.os.WriteBit(opcodeUpdateSig)
+		if sig == 0 {
+			enc.os.WriteBit(opcodeZeroSig)
+		} else {
+			enc.os.WriteBit(opcodeNonZeroSig)
+			enc.os.WriteBits(uint64(sig-1), numSigBits)
+		}
+
+		enc.numSig = sig
+	} else {
+		enc.os.WriteBit(opcodeNoUpdate)
+	}
+
+	if mult > enc.maxMult {
+		enc.os.WriteBit(opcodeUpdateMult)
+		enc.os.WriteBits(uint64(mult), numMultBits)
+		enc.maxMult = mult
+	} else {
+		enc.os.WriteBit(opcodeNoUpdate)
+	}
+}
+
+// getNewSig gets the new number of significant bits given the
+// number of significant bits of the current diff. It takes into
+// account thresholds to try and find a value that's best for the
+// current data
+func (enc *encoder) getNewSig(numSig uint8) uint8 {
+	newSig := enc.numSig
+
+	if numSig > enc.numSig {
+		newSig = numSig
+	} else if enc.numSig-numSig >= sigDiffThreshold {
+		if enc.numLowerSig == 0 {
+			enc.curLowestSig = numSig
+		} else if numSig > enc.curLowestSig {
+			enc.curLowestSig = numSig
+		}
+
+		enc.numLowerSig++
+		if enc.numLowerSig >= sigRepeatThreshold {
+			newSig = enc.curLowestSig
+			enc.numLowerSig = 0
+		}
+
+	} else {
+		enc.numLowerSig = 0
+	}
+
+	return newSig
```

Comment:
Hm, it's just weird that it mutates some encoder state but not others.  Perhaps we just rename this to `accountAndGetNewSig(...)` to reflect that it's doing some accounting as a side effect of calling it?